migrator/replica/postgres/db_replica_manager_impl.go

```diff
@@ -0,0 +1,297 @@
+	// If the versions are the same, but rocks has a more recent update then we need to migrate rocks to postgres
+	// Otherwise we roll with Postgres->Postgres
+	if rocksVersion != nil && !rocksVersion.LastPersisted.IsZero() && currReplica.GetMigVersion().LastPersisted.Before(rocksVersion.LastPersisted) {
+		// Create a temp replica for processing of current
+		// Seed it from an empty database because we need to run migrations from Rocks to Postgres.
+		if !d.databaseExists(TempReplica) {
+			err := pgadmin.CreateDB(d.sourceMap, d.adminConfig, pgadmin.EmptyDB, TempReplica)
+
+			// If for some reason, we cannot create a temp replica we will need to continue to upgrade
+			// with the current and thus no fallback.
+			if err != nil {
+				log.Errorf("Unable to create Temp database: %v", err)
+			}
+		}
+		d.ReplicaMap[TempReplica] = metadata.NewPostgres(d.ReplicaMap[CurrentReplica].GetMigVersion(), TempReplica)
+		return TempReplica, true, nil
+	}
+
+	prevReplica, prevExists := d.ReplicaMap[PreviousReplica]
+	if d.rollbackEnabled() && currReplica.GetVersion() != version.GetMainVersion() {
+		// If previous replica has the same version as current version, the previous upgrade was not completed.
+		// Central could be in a loop of booting up the service. So we should continue to run with current.
+		if prevExists && currReplica.GetVersion() == prevReplica.GetVersion() {
+			return CurrentReplica, false, nil
+		}
+		if version.CompareVersions(currReplica.GetVersion(), version.GetMainVersion()) > 0 || currReplica.GetSeqNum() > migrations.CurrentDBVersionSeqNum() {
+			// Force rollback
+			return PreviousReplica, false, nil
+		}
+
+		d.safeRemove(PreviousReplica)
+		if d.hasSpaceForRollback() {
+			// Create a temp replica for processing of current
+			// If such a replica already exists then we were previously in the middle of processing
+			if !d.databaseExists(TempReplica) {
+				err := pgadmin.CreateDB(d.sourceMap, d.adminConfig, CurrentReplica, TempReplica)
+
+				// If for some reason, we cannot create a temp replica we will need to continue to upgrade
+				// with the current and thus no fallback.
+				if err != nil {
+					log.Errorf("Unable to create Temp database: %v", err)
+				}
+			}
+			d.ReplicaMap[TempReplica] = metadata.NewPostgres(d.ReplicaMap[CurrentReplica].GetMigVersion(), TempReplica)
+			return TempReplica, false, nil
+
+		}
+
+		// If the space is not enough to make a replica, continue to upgrade with current.
+		return CurrentReplica, false, nil
+	}
+
+	// Rollback from previous version.
+	if prevExists && prevReplica.GetVersion() == version.GetMainVersion() {
+		return PreviousReplica, false, nil
+	}
+
+	log.Info("Fell through all checks to return current, meaning probably empty OR rollback disabled.")
+	return CurrentReplica, false, nil
+}
+
+// Persist - replaces current replica with upgraded one.
+func (d *DBReplicaManagerImpl) Persist(replicaName string) error {
```

Comment:
It actually does, if rollback is enabled, all work will be done on Temp Replica and the act of persisting that will do that.  Now one thing we might want to consider in this situation is what does enabling rollback mean in this case?  Do we use different criteria to determine rollback if we are in this hybrid situation or do we follow the same rules?  Since rollback is based on space we would probably have to do something sneaky like move current to previous and have an empty current.  But it seemed simpler to just not allow rollback from postgres->rocks if rollback from rocks->rocks was not enabled.